### PR TITLE
PMM-9459 Remove usage of DescribeByCollect

### PR DESCRIPTION
**Pull request by ShashankSinha252**
DescribeByCollect method internally uses Collect method. This causes exporter to retrieve metric data from database twice per scrape. This patch reduces monitoring load on database server, as well as reduces CPU and memory usage of exporter.

[PMM-9459](https://jira.percona.com/browse/PMM-9459)

### CODE REVIEW

**Comment by ShashankSinha252** [comment]
This change adds an assumption on usage of collectors. `Describe` and `Collect` methods are interlinked, since Describe creates the metric cache consumed by Collect. For every call to `Collect`, we need to ensure a call to `Describe` before it.

**Comment by BupycHuk** [comment]
Can we extract duplicated code to a separate structure?

File: exporter/collstats_collector.go [hunk 1]
```diff
@@ -27,20 +28,63 @@ import (
...
 func (d *collstatsCollector) Collect(ch chan<- prometheus.Metric) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	for _, metric := range d.metricsCache {
+		ch <- metric
+	}
+}
+
+func (d *collstatsCollector) collect(ch chan<- prometheus.Metric) {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go:60]: Will be ignored as there is no code change here.

File: exporter/collstats_collector.go [hunk 2]
```diff
@@ -27,20 +28,63 @@ import (
 )
 
 type collstatsCollector struct {
-	ctx             context.Context
-	client          *mongo.Client
-	collections     []string
+	ctx    context.Context
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go]: Will be ignored as there is no code change here.

File: exporter/diagnostic_data_collector.go [hunk 3]
```diff
@@ -27,18 +28,58 @@ import (
 )
 
 type diagnosticDataCollector struct {
-	ctx            context.Context
-	client         *mongo.Client
+	ctx    context.Context
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/diagnostic_data_collector.go]: Will be ignored as there is no code change here.

File: exporter/collstats_collector.go [hunk 4]
```diff
@@ -21,32 +21,56 @@ import (
...
 func (d *collstatsCollector) Describe(ch chan<- *prometheus.Desc) {
-	prometheus.DescribeByCollect(d, ch)
+	d.base.Describe(ch, d.collect)
 }
 
 func (d *collstatsCollector) Collect(ch chan<- prometheus.Metric) {
+	d.base.Collect(ch)
+}
+
+func (d *collstatsCollector) collect(ch chan<- prometheus.Metric) {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go:60]: Will be ignored as there is no code change here.

File: exporter/indexstats_collector.go [hunk 5]
```diff
@@ -22,31 +22,53 @@ import (
...
 func (d *indexstatsCollector) Describe(ch chan<- *prometheus.Desc) {
-	prometheus.DescribeByCollect(d, ch)
+	d.base.Describe(ch, d.collect)
 }
 
 func (d *indexstatsCollector) Collect(ch chan<- prometheus.Metric) {
+	d.base.Collect(ch)
+}
+
+func (d *indexstatsCollector) collect(ch chan<- prometheus.Metric) {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/indexstats_collector.go:58]: Will be ignored as there is no code change here.

File: exporter/collstats_collector.go [hunk 6]
```diff
@@ -21,32 +21,56 @@ import (
...
-	"github.com/sirupsen/logrus"
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/mongo"
 )
 
 type collstatsCollector struct {
-	ctx             context.Context
-	client          *mongo.Client
-	collections     []string
+	ctx  context.Context
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go:29]: Will be ignored as there is no code change here.

File: exporter/dbstats_collector.go [hunk 7]
```diff
@@ -20,46 +20,68 @@ import (
...
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/sirupsen/logrus"
 	"go.mongodb.org/mongo-driver/bson"
-	"go.mongodb.org/mongo-driver/mongo"
 )
 
 type dbstatsCollector struct {
-	ctx            context.Context
-	client         *mongo.Client
+	ctx  context.Context
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/dbstats_collector.go:27]: Will be ignored as there is no code change here.

File: exporter/diagnostic_data_collector.go [hunk 8]
```diff
@@ -21,66 +21,85 @@ import (
...
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/sirupsen/logrus"
 	"go.mongodb.org/mongo-driver/bson"
-	"go.mongodb.org/mongo-driver/mongo"
 )
 
 type diagnosticDataCollector struct {
-	ctx            context.Context
-	client         *mongo.Client
+	ctx  context.Context
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/diagnostic_data_collector.go:28]: Will be ignored as there is no code change here.

File: exporter/diagnostic_data_collector.go [hunk 9]
```diff
@@ -21,66 +21,85 @@ import (
...
 func (d *diagnosticDataCollector) Describe(ch chan<- *prometheus.Desc) {
-	prometheus.DescribeByCollect(d, ch)
+	d.base.Describe(ch, d.collect)
 }
 
 func (d *diagnosticDataCollector) Collect(ch chan<- prometheus.Metric) {
+	d.base.Collect(ch)
+}
+
+func (d *diagnosticDataCollector) collect(ch chan<- prometheus.Metric) {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/diagnostic_data_collector.go:53]: Will be ignored as there is no code change here.

File: exporter/collstats_collector.go [hunk 10]
```diff
@@ -27,20 +28,63 @@ import (
...
+
 	compatibleMode  bool
 	discoveringMode bool
-	logger          *logrus.Logger
 	topologyInfo    labelsGetter
+
+	collections []string
+}
+
+func NewCollectionStatsCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, compatible, discovery bool, topology labelsGetter, collections []string) *collstatsCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go]: Unexported

File: exporter/dbstats_collector.go [hunk 11]
```diff
@@ -26,20 +27,62 @@ import (
...
+	metricsCache []prometheus.Metric
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
+
+	databaseFilter []string
+}
+
+func NewDBStatsCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, compatible bool, topology labelsGetter, databaseRegex []string) *dbstatsCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/dbstats_collector.go]: Unexported

File: exporter/diagnostic_data_collector.go [hunk 12]
```diff
@@ -27,18 +28,58 @@ import (
...
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewDiagnosticDataCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, compatible bool, topology labelsGetter) *diagnosticDataCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/diagnostic_data_collector.go]: Unexported

File: exporter/general_collector.go [hunk 13]
```diff
@@ -31,13 +32,50 @@ type generalCollector struct {
 	ctx    context.Context
 	client *mongo.Client
 	logger *logrus.Logger
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+}
+
+func NewGeneralCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger) *generalCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/general_collector.go]: Unexported

File: exporter/indexstats_collector.go [hunk 14]
```diff
@@ -28,19 +29,61 @@ import (
...
+	metricsCache []prometheus.Metric
+
 	discoveringMode bool
-	logger          *logrus.Logger
 	topologyInfo    labelsGetter
+
+	collections []string
+}
+
+func NewIndexStatsCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, discovery bool, topology labelsGetter, collections []string) *indexstatsCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/indexstats_collector.go]: Unexported

File: exporter/replset_status_collector.go [hunk 15]
```diff
@@ -31,18 +32,58 @@ const (
...
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewReplicationSetStatusCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, compatible bool, topology labelsGetter) *replSetGetStatusCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/replset_status_collector.go]: Unexported

File: exporter/serverstatus_collector.go [hunk 16]
```diff
@@ -26,18 +27,58 @@ import (
...
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewServerStatusCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, compatible bool, topology labelsGetter) *serverStatusCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/serverstatus_collector.go]: Unexported

File: exporter/top_collector.go [hunk 17]
```diff
@@ -26,18 +27,59 @@ import (
...
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewTopCollector(ctx context.Context, client *mongo.Client, logger *logrus.Logger, compatible bool, topology labelsGetter) *topCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/top_collector.go]: Unexported

File: exporter/dbstats_collector.go [hunk 18]
```diff
@@ -26,20 +27,62 @@ import (
 )
 
 type dbstatsCollector struct {
-	ctx            context.Context
-	client         *mongo.Client
+	ctx    context.Context
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/dbstats_collector.go]: Will be ignored as there is no code change here.

File: exporter/diagnostic_data_collector.go [hunk 19]
```diff
@@ -27,18 +28,58 @@ import (
...
 func (d *diagnosticDataCollector) Collect(ch chan<- prometheus.Metric) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	for _, metric := range d.metricsCache {
+		ch <- metric
+	}
+}
+
+func (d *diagnosticDataCollector) collect(ch chan<- prometheus.Metric) {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/diagnostic_data_collector.go:54]: Will be ignored as there is no code change here.

File: exporter/top_collector.go [hunk 20]
```diff
@@ -22,24 +22,44 @@ import (
...
-	client         *mongo.Client
+	ctx  context.Context
+	base *baseCollector
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewTopCollector(ctx context.Context, base *baseCollector, compatible bool, topology labelsGetter) *topCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/top_collector.go]: Unexported

File: exporter/dbstats_collector.go [hunk 21]
```diff
@@ -20,46 +20,68 @@ import (
...
+	base *baseCollector
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
+
+	databaseFilter []string
+}
+
+func NewDBStatsCollector(ctx context.Context, base *baseCollector, compatible bool, topology labelsGetter, databaseRegex []string) *dbstatsCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/dbstats_collector.go]: Unexported

File: exporter/diagnostic_data_collector.go [hunk 22]
```diff
@@ -21,66 +21,85 @@ import (
...
-	client         *mongo.Client
+	ctx  context.Context
+	base *baseCollector
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewDiagnosticDataCollector(ctx context.Context, base *baseCollector, compatible bool, topology labelsGetter) *diagnosticDataCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/diagnostic_data_collector.go]: Unexported

File: exporter/general_collector.go [hunk 23]
```diff
@@ -28,17 +28,30 @@ import (
...
 // command to gather stats.
 type generalCollector struct {
-	ctx    context.Context
-	client *mongo.Client
-	logger *logrus.Logger
+	ctx  context.Context
+	base *baseCollector
+}
+
+func NewGeneralCollector(ctx context.Context, base *baseCollector) *generalCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/general_collector.go]: Unexported

File: exporter/indexstats_collector.go [hunk 24]
```diff
@@ -22,31 +22,53 @@ import (
...
+	base *baseCollector
+
 	discoveringMode bool
-	logger          *logrus.Logger
 	topologyInfo    labelsGetter
+
+	collections []string
+}
+
+func NewIndexStatsCollector(ctx context.Context, base *baseCollector, discovery bool, topology labelsGetter, collections []string) *indexstatsCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/indexstats_collector.go]: Unexported

File: exporter/replset_status_collector.go [hunk 25]
```diff
@@ -31,20 +30,41 @@ const (
...
-	client         *mongo.Client
+	ctx  context.Context
+	base *baseCollector
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewReplicationSetStatusCollector(ctx context.Context, base *baseCollector, compatible bool, topology labelsGetter) *replSetGetStatusCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/replset_status_collector.go]: Unexported

File: exporter/serverstatus_collector.go [hunk 26]
```diff
@@ -22,24 +22,43 @@ import (
...
-	client         *mongo.Client
+	ctx  context.Context
+	base *baseCollector
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
 }
 
+func NewServerStatusCollector(ctx context.Context, base *baseCollector, compatible bool, topology labelsGetter) *serverStatusCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/serverstatus_collector.go]: Unexported

File: exporter/collstats_collector.go [hunk 27]
```diff
@@ -21,32 +21,56 @@ import (
...
+
 	compatibleMode  bool
 	discoveringMode bool
-	logger          *logrus.Logger
 	topologyInfo    labelsGetter
+
+	collections []string
+}
+
+func NewCollectionStatsCollector(ctx context.Context, base *baseCollector, compatible, discovery bool, topology labelsGetter, collections []string) *collstatsCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go]: Unexported

File: exporter/base_collector.go [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+
+type baseCollector struct {
+	client *mongo.Client
+	logger *logrus.Logger
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+}
+
+func NewBaseCollector(client *mongo.Client, logger *logrus.Logger) *baseCollector {
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/base_collector.go]: Unexported

File: exporter/base_collector.go [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+
+type baseCollector struct {
+	client *mongo.Client
+	logger *logrus.Logger
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+}
+
+// newBaseCollector creates a skeletal collector, which is used to create other collectors
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/base_collector.go]: Resolved

File: exporter/collstats_collector.go [hunk 30]
```diff
@@ -21,32 +21,57 @@ import (
...
+
 	compatibleMode  bool
 	discoveringMode bool
-	logger          *logrus.Logger
 	topologyInfo    labelsGetter
+
+	collections []string
+}
+
+// newCollectionStatsCollector creates a collector for statistics about collections
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go]: Resolved

File: exporter/dbstats_collector.go [hunk 31]
```diff
@@ -20,46 +20,69 @@ import (
...
+	base *baseCollector
+
 	compatibleMode bool
-	logger         *logrus.Logger
 	topologyInfo   labelsGetter
+
+	databaseFilter []string
+}
+
+// newDBStatsCollector creates a collector for statistics on database storage
```
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/dbstats_collector.go]: Resolved

File: exporter/base_collector.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/base_collector.go]: Resolved

File: exporter/collstats_collector.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/collstats_collector.go]: Resolved

File: exporter/dbstats_collector.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (ShashankSinha252) [reply exporter/dbstats_collector.go]: Resolved

File: exporter/base_collector.go [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+
+	lock         sync.Mutex
+	metricsCache []prometheus.Metric
+}
+
+// newBaseCollector creates a skeletal collector, which is used to create other collectors.
+func newBaseCollector(client *mongo.Client, logger *logrus.Logger) *baseCollector {
+	return &baseCollector{
+		client: client,
+		logger: logger,
```
Comment (percona-csalguero) [inline exporter/base_collector.go:39]:
```suggestion
		logger: logger,
		lock: &sync.Mutex{},
```

File: exporter/exporter.go [hunk 33]
```diff
@@ -158,74 +160,42 @@ func (e *Exporter) makeRegistry(ctx context.Context, client *mongo.Client, topol
...
-			ctx:             ctx,
-			client:          client,
-			collections:     e.opts.CollStatsNamespaces,
-			compatibleMode:  e.opts.CompatibleMode,
-			discoveringMode: e.opts.DiscoveringMode,
-			logger:          e.opts.Logger,
-			topologyInfo:    topologyInfo,
-		}
-		registry.MustRegister(&cc)
+		cc := newCollectionStatsCollector(ctx, base,
```
Comment (percona-csalguero) [inline exporter/exporter.go:163]: We are adding collectors like this one (StatsCollector) to the the registry and Prometheus then will call each collector on every scrape. By having one single base instance and passing it to every collector we are making them to share the lock and the metrics cache. 
AFAIK, Prometheus call each collector sequentially but if they change this behavior and call collectors concurrently each collector will lock the others. 
What you think about call the `newXXXCollector` with `client` and any other parameter and create a new and independent base for each collector?

**Review by percona-csalguero — CHANGES_REQUESTED** [review]

File: exporter/base_collector.go (hunk 32 above)
Comment (BupycHuk) [reply exporter/base_collector.go:39]: mutex doesn't have to be a pointer

File: exporter/exporter.go (hunk 33 above)
Comment (BupycHuk) [reply exporter/exporter.go:163]: +1

File: exporter/diagnostic_data_collector.go [hunk 34]
```diff
@@ -21,66 +21,86 @@ import (
...
-	"github.com/sirupsen/logrus"
 	"go.mongodb.org/mongo-driver/bson"
-	"go.mongodb.org/mongo-driver/mongo"
 )
 
 type diagnosticDataCollector struct {
-	ctx            context.Context
-	client         *mongo.Client
+	ctx  context.Context
+	base *baseCollector
```
Comment (BupycHuk) [inline exporter/diagnostic_data_collector.go:29]: I'm not sure that base should be a pointer.

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Agree with @percona-csalguero about sharing the same base structure across different collectors.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

**Review by percona-csalguero — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Review by artemgavrilov — APPROVED** [review]

**Pull request merged**